analyzer/src/main/kotlin/managers/Cargo.kt

```diff
@@ -0,0 +1,290 @@
+
+    private fun extractVcsInfo(node: JsonNode): VcsInfo {
+        val url = extractRepositoryUrl(node)
+        val type = if (url != null) "git" else "" // for now cargo supports only git
+        return VcsInfo(type, url.orEmpty(), revision = "")
+    }
+
+    private fun extractDeclaredLicenses(node: JsonNode): SortedSet<String> {
+        return node["license"].textValueOrEmpty().split("/")
+            .map { it.trim() }
+            .filter { it.isNotBlank() }
+            .toSortedSet()
+    }
+
+    private fun extractSourceArtifact(
+        node: JsonNode,
+        hashes: Map<String, String>
+    ): RemoteArtifact? {
+        if (node["source"].textValueOrEmpty()
+            == "registry+https://github.com/rust-lang/crates.io-index"
+        ) {
+            val name = node["name"]?.asText() ?: return null
+            val version = node["version"]?.asText() ?: return null
+            val url = "https://crates.io/api/v1/crates/$name/$version/download"
+            val checksum = checksumKeyOf(node)
+            val hash = Hash.create(hashes[checksum] ?: "")
+            return RemoteArtifact(url, hash)
+        } else {
+            return null
+        }
+    }
+
+    private fun extractPackage(node: JsonNode, hashes: Map<String, String>): Package {
+        val vcsInfo = extractVcsInfo(node)
+        return Package(
+            id = extractPackageId(node),
+            declaredLicenses = extractDeclaredLicenses(node),
+            description = node["description"].textValueOrEmpty(),
+            binaryArtifact = RemoteArtifact.EMPTY,
+            sourceArtifact = extractSourceArtifact(node, hashes) ?: RemoteArtifact.EMPTY,
+            homepageUrl = "",
+            vcs = vcsInfo,
+            vcsProcessed = vcsInfo.normalize()
+        )
+    }
+
+    private fun checksumKeyOf(metadata: JsonNode): String {
+        val id = metadata["id"]!!.textValue()
+        return "\"checksum $id\""
+    }
+
+    // Cargo.lock is located next to Cargo.toml or in one of the parent directories. The latter
+    // is case when the project is part of a workspace. Cargo.lock is then located next to the
+    // Cargo.toml file defining the workspace.
+    private fun findCargoLock(definitionFile: File): File {
+        var workingDir = definitionFile.parentFile
+        while (workingDir != analyzerRoot.parentFile) {
+            val lockfile = File(workingDir, "Cargo.lock")
+            if (lockfile.exists()) {
+                return lockfile
+            }
+            workingDir = workingDir.parentFile
+        }
+        // we reached the root directory we are analyzing
+        throw IllegalArgumentException("missing Cargo.lock file")
+    }
+
+    private fun readHashes(lockfile: File): Map<String, String> {
+        val contents = Toml().read(lockfile)
+        val metadata = contents.getTable("metadata") ?: return emptyMap()
+        val metadataMap = metadata.toMap()
+        return metadataMap.asSequence().mapNotNull {
```

Comment:
The view as a sequence doe snot serve any purpose here, or? How about this instead:

        val metadataMapNotNull = mutableMapOf<String, String>()

        metadataMap.forEach { (key, value) ->
            if (key != null && (value as? String) != null) {
                metadataMapNotNull[key] = value
            }
        }

        return metadataMapNotNull

Reply:
not very functional, but ok